File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        def dfs(cur_node):
            if not cur_node:
                return 0
            
            return max(dfs(cur_node.left) + 1, dfs(cur_node.right) + 1)
        
        return dfs(root)

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int res = 0;
    int dfs(TreeNode* curNode){
        if (curNode == nullptr){
            return 0;
        }
        int leftDepth = dfs(curNode->left);
        int rightDepth = dfs(curNode->right);
        res = max(res, leftDepth + rightDepth);
        return 1 + max(rightDepth, leftDepth);
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int tmpRes = dfs(root);
        return res;
    }
};
